File: lib/py/uapi/internal/types/UGeneric.py
from typing import List, Dict, TYPE_CHECKING

if TYPE_CHECKING:
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
    from uapi.internal.validation.ValidationFailure import ValidationFailure

from uapi.internal.types.UType import UType


class UGeneric(UType):
    def __init__(self, index: int):
        self.index = index

    def get_type_parameter_count(self) -> int:
        return 0

    def validate(self, value: object, select: Dict[str, object], fn: str,
                 type_parameters: List['UTypeDeclaration'], generics: List['UTypeDeclaration']) -> List['ValidationFailure']:
        type_declaration = generics[self.index]
        return type_declaration.validate(value, select, fn, [])

    def generate_random_value(self, blueprint_value: object, use_blueprint_value: bool,
                              include_optional_fields: bool, randomize_optional_fields: bool,
                              type_parameters: List['UTypeDeclaration'], generics: List['UTypeDeclaration'],
                              random_generator: 'RandomGenerator') -> object:
        generic_type_declaration = generics[self.index]
        return generic_type_declaration.generate_random_value(blueprint_value, use_blueprint_value,
                                                              include_optional_fields, randomize_optional_fields,
                                                              [], random_generator)

    def get_name(self, generics: List['UTypeDeclaration']) -> str:
        type_declaration = generics[self.index]
        return type_declaration.type.get_name(generics)

File: lib/ts/uapi/internal/types/UGeneric.ts
import { RandomGenerator } from 'uapi/RandomGenerator';
import { UTypeDeclaration } from 'uapi/internal/types/UTypeDeclaration';
import { ValidationFailure } from 'uapi/internal/validation/ValidationFailure';
import { UType } from 'uapi/internal/types/UType';

export class UGeneric extends UType {
    private index: number;

    constructor(index: number) {
        super();
        this.index = index;
    }

    get_type_parameter_count(): number {
        return 0;
    }

    validate(
        value: any,
        select: { [key: string]: any } | null,
        fn: string | null,
        type_parameters: UTypeDeclaration[],
        generics: UTypeDeclaration[],
    ): ValidationFailure[] {
        const type_declaration = generics[this.index];
        return type_declaration.validate(value, select, fn, []);
    }

    generate_random_value(
        blueprint_value: any,
        use_blueprint_value: boolean,
        include_optional_fields: boolean,
        randomize_optional_fields: boolean,
        type_parameters: UTypeDeclaration[],
        generics: UTypeDeclaration[],
        random_generator: RandomGenerator,
    ): any {
        const generic_type_declaration = generics[this.index];
        return generic_type_declaration.generate_random_value(
            blueprint_value,
            use_blueprint_value,
            include_optional_fields,
            randomize_optional_fields,
            [],
            random_generator,
        );
    }

    get_name(generics: UTypeDeclaration[]): string {
        const type_declaration = generics[this.index];
        return type_declaration.type.get_name(generics);
    }
}
